Kenneth-Macharia/FastFoodFast
# #163191061 add user login and logout features

#### What does this PR do?
- Adds JWT authentication to the app
#### Description of Tasks to be completed
- Using JWT adds user login feature on the endpoint '/v1/auth/login'
- Using JWT token blacklisting, adds user logout feature on the endpoint '/v1/auth/logout'
- Adds user rights control i.e. a default "Guest' user cannot access 'Admin' routes endpoints like the 
 ones dealing with menu management e.g '/menus' for retrieving all menu items.
#### How should this be manually tested?
##### Database setup
- Install PostgreSQL on a local machine
- Create a 'MAIN' database on that machine
- Create a local user to the database with full rights to it, secured by a password
##### App setup and running
- Clone this repo to the local machine with the database above
- Navigate to the root of the repo: '/FastFoodFast
- Checkout the branch: 'ft-auth-login_POST-163191061'
- Create a virtual environment > virtualenv <your_environment_name>
- Activate the virtual environment and install all app requirements: pip install -r requirements.txt
- Set/export the the environment variable specified in the .env file with your local details
- Run the app: 'python run.py' at the command prompt.
- The app server should run and give the URL at which it is listening for requests
##### Testing the app in Postman
- Install postman on the local machine
- Create a user by performing a POST request to the '/v1/auth/signup' route with the following payload:
{"User_Email":"<random email>", "User_Password":<"random password>", "User_Name":<random name"}. The response "Successfully signed up should be returned.
- Attempt a GET to the endpoint '/menus, the response "Missing authorization header" meaning a logged in token has not been furnished for this request.
- Login the above created user via a POST to the endpoint '/v1/auth/login' and furnish the email and password details above as the payload. The response "Successfully signed in should be returned", and a token generated.
- Attempt the GET above to the menus endpoint above but add the an authorization key in the request header and in the value field input "Bearer <the-token-generated-above>" before sending the request. The response "This is an admin only function" should appear, meaning the user is now logged but can access this endpoint, since he is a 'Guest'.

#### What are the relevant pivotal tracker stories?
#163191061
